packages/atom/src/atom.ts

```diff
@@ -38,8 +38,18 @@ export class Atom<S> implements Subscribable<S,undefined> {
     }
   }
 
-  slice<T>(path: string[]): Slice<T, S> {
-    return new Slice(this, path);
+
+  slice<Key extends keyof S>(key: Key): Slice<S[Key], S>;
+  slice<Key1 extends keyof S, Key2 extends keyof S[Key1]>(key1: Key1, key2: Key2): Slice<S[Key1][Key2], S>;
+  slice<Key1 extends keyof S, Key2 extends keyof S[Key1], Key3 extends keyof S[Key1][Key2]>(key1: Key1, key2: Key2, key3: Key3): Slice<S[Key1][Key2][Key3], S>;
+  slice<Key1 extends keyof S, Key2 extends keyof S[Key1], Key3 extends keyof S[Key1][Key2], Key4 extends keyof S[Key1][Key2][Key3]>(key1: Key1, key2: Key2, key3: Key3, key4: Key4): Slice<S[Key1][Key2][Key3][Key4], S>;
+  slice<Key1 extends keyof S, Key2 extends keyof S[Key1], Key3 extends keyof S[Key1][Key2], Key4 extends keyof S[Key1][Key2][Key3], Key5 extends keyof S[Key1][Key2][Key3][Key4]>(key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5): Slice<S[Key1][Key2][Key3][Key4][Key5], S>;
+  slice<Key1 extends keyof S, Key2 extends keyof S[Key1], Key3 extends keyof S[Key1][Key2], Key4 extends keyof S[Key1][Key2][Key3], Key5 extends keyof S[Key1][Key2][Key3][Key4], Key6 extends keyof S[Key1][Key2][Key3][Key4][Key5]>(key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5, key6: Key6): Slice<S[Key1][Key2][Key3][Key4][Key5][Key6], S>;
+  slice<Key1 extends keyof S, Key2 extends keyof S[Key1], Key3 extends keyof S[Key1][Key2], Key4 extends keyof S[Key1][Key2][Key3], Key5 extends keyof S[Key1][Key2][Key3][Key4], Key6 extends keyof S[Key1][Key2][Key3][Key4][Key5], Key7 extends keyof S[Key1][Key2][Key3][Key4][Key5][Key6]>(key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5, key6: Key6, key7: Key7): Slice<S[Key1][Key2][Key3][Key4][Key5][Key6][Key7], S>;
+  slice<Key1 extends keyof S, Key2 extends keyof S[Key1], Key3 extends keyof S[Key1][Key2], Key4 extends keyof S[Key1][Key2][Key3], Key5 extends keyof S[Key1][Key2][Key3][Key4], Key6 extends keyof S[Key1][Key2][Key3][Key4][Key5], Key7 extends keyof S[Key1][Key2][Key3][Key4][Key5][Key6], Key8 extends keyof S[Key1][Key2][Key3][Key4][Key5][Key6][Key7]>(key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5, key6: Key6, key7: Key7, key8: Key8): Slice<S[Key1][Key2][Key3][Key4][Key5][Key6][Key7][Key8], S>;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  slice(...keys: string[]): Slice<any, S> {
+    return new Slice(this, keys);
```

Comment:
Is there a way to put all these overrides into an interface so that they can not only be shared between Atom and Slice, but so that they can also be just put into a different file? :)

Reply:
I've tried a few things but and creating a generic function

```typescript

export function slice<A, S, Key extends keyof A>(atom: A, key: Key): Slice<A[Key], S>;
export function slice<A, S, Key1 extends keyof A, Key2 extends keyof A[Key1]>(atom: A, key1: Key1, key2: Key2): Slice<A[Key1][Key2], S>;
export function slice<A, S, Key1 extends keyof A, Key2 extends keyof A[Key1], Key3 extends keyof A[Key1][Key2]>(atom: A, key1: Key1, key2: Key2, key3: Key3): Slice<A[Key1][Key2][Key3], S>;
export function slice<A, S, Key1 extends keyof A, Key2 extends keyof A[Key1], Key3 extends keyof A[Key1][Key2], Key4 extends keyof A[Key1][Key2][Key3]>(atom: A, key1: Key1, key2: Key2, key3: Key3, key4: Key4): Slice<A[Key1][Key2][Key3][Key4], S>;
export function slice<A, S, Key1 extends keyof A, Key2 extends keyof A[Key1], Key3 extends keyof A[Key1][Key2], Key4 extends keyof A[Key1][Key2][Key3], Key5 extends keyof A[Key1][Key2][Key3][Key4]>(atom: A, key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5): Slice<A[Key1][Key2][Key3][Key4][Key5], S>;
export function slice<A, S, Key1 extends keyof A, Key2 extends keyof A[Key1], Key3 extends keyof A[Key1][Key2], Key4 extends keyof A[Key1][Key2][Key3], Key5 extends keyof A[Key1][Key2][Key3][Key4], Key6 extends keyof A[Key1][Key2][Key3][Key4][Key5]>(atom: A, key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5, key: Key6): Slice<A[Key1][Key2][Key3][Key4][Key5][Key6], S>;
export function slice<A, S, Key1 extends keyof A, Key2 extends keyof A[Key1], Key3 extends keyof A[Key1][Key2], Key4 extends keyof A[Key1][Key2][Key3], Key5 extends keyof A[Key1][Key2][Key3][Key4], Key6 extends keyof A[Key1][Key2][Key3][Key4][Key5], Key7 extends keyof A[Key1][Key2][Key3][Key4][Key5][Key6]>(atom: A, key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5, key: Key6, key7: Key7): Slice<A[Key1][Key2][Key3][Key4][Key5][Key6][Key7], S>;
export function slice<A, S, Key1 extends keyof A, Key2 extends keyof A[Key1], Key3 extends keyof A[Key1][Key2], Key4 extends keyof A[Key1][Key2][Key3], Key5 extends keyof A[Key1][Key2][Key3][Key4], Key6 extends keyof A[Key1][Key2][Key3][Key4][Key5], Key7 extends keyof A[Key1][Key2][Key3][Key4][Key5][Key6], Key8 extends keyof A[Key1][Key2][Key3][Key4][Key5][Key6][Key7]>(atom: A, key1: Key1, key2: Key2, key3: Key3, key4: Key4, key5: Key5, key: Key6, key7: Key7, key8: Key8): Slice<A[Key1][Key2][Key3][Key4][Key5][Key6][Key7][Key8], S>;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function slice<A, S>(atom: Atom<A>, keys: Key[]): Slice<S, A> {
    return new Slice(atom, keys);
  }
```

and then from the atom and slice functions but the generics get lost.

Reply:
ok, in that case, let's roll with it so that we don't get bit by type errors anymore.